book/src/pil/types.md

```diff
@@ -0,0 +1,249 @@
+- `|`: bit-wise disjunction
+- `^`: bit-wise exclusive or
+- `<<`: bit-wise shift left, the shift amount needs to be non-negative and fit 32 bits, otherwise a runtime error is triggered
+- `>>`: bit-wise shift right, the shift amount needs to be non-negative and fit 32 bits, otherwise a runtime error is triggered
+
+The exponentiation operator on field elements requires a non-negative integer as exponent.
+It has the signature `**: fe, int -> fe`.
+
+In addition, the following comparison operators are allowed, the result is a boolean:
+- `<`: less than
+- `<=`: less or equal
+- `==`: equal
+- `!=`: not equal
+- `>=`: greater or equal
+- `>`: greater than
+
+### Field Element
+
+Type name: `fe`
+
+Field elements are elements of a particular but unspecified prime field. The exact field is
+chosen when powdr is run. The modulus of that field can be accessed via `std::field::modulus()`.
+
+Field elements are the values stored in (fixed, witness and intermediate) columns.
+Arithmetic inside constraints (algebraic expressions) is also always finite field arithmetic.
+
+Since literal numbers in the source code are integers, those have to be converted to field elements
+using the `std::convert::fe` function. In some situations (see below), integers
+(including literal numbers) are converted to field elements implicitly.
+
+Field elements allow the following operators, where the result is always a field element:
+
+- `+`: finite field addition
+- `-`: finite field subtraction (also unary negation)
+- `*`: finite field multiplication
+
+There is also an exponentiation operator `**` on field elements. It requires the exponent
+to be a non-negative integer and thus has the signature `**: fe, int -> fe`. If the exponent is negative, a runtime error is triggered. `0**0` is defined as `1`.
+
+The following comparison operators exist for field elements, whose result is a boolean:
+
+- `==`: equality comparison
+- `!=`: inequality comparison
+
+Since finite fields do not have an inherent order as integers do, if you want to
+compare them using `<`, you have to first convert them to integers.
+
+### String
+
+Type name: `string`
+
+String literals are written as `"string content"`. They are mainly used for debugging or
+documentation purposes, since they cannot occur in constraints.
+
+They allow the following operators:
+
+- `+`: string concatenation
+
+### Tuple
+
+Type name: `(..., ..., ...)`
+
+Tuples are complex types that are composed from other types, either zero or two or more. There is no tuple type with a single element (`(int)` is the same as `int`).
+The empty tuple type is written as `()`.
+
+Examples include `(int, int)` (a pair of integers) and `((fe[], int), ())`
+(a tuple consisting of a tuple that contains an array of field elements and an integer
+and an empty tuple).
+
+Tuples values are constructed using parentheses: `(1, 2)` constructs the tuple that consists of
+a one and a two.
+
+Tuples do not allow any operators.
```

Comment:
Not even indexing? Can tuples be destructed? How would I use them?

Reply:
Indeed, they are totally useless currently. You would need to destructure them, but that is not yet implemented. We could support `.0`, but we also don't have member access yet.